tests/spsc/ring_buf.cpp

```diff
@@ -319,6 +318,36 @@ TEST_CASE("spsc::RingBuf - Peek std::array", "[rb_peek_std_array]") {
         std::equal(test_data.begin(), test_data.end(), test_data_read.begin()));
 }
 
+TEST_CASE("Get available after clear", "[rb_get_available_after_clear]") {
+    lockfree::spsc::RingBuf<uint64_t, 1024U> rb;
+    const std::array<uint64_t, 512> test_data = {0xE5U};
+
+    rb.Write(test_data);
+    rb.Clear();
+
+    REQUIRE(rb.GetAvailable() == 0);
+}
+
+TEST_CASE("Get free after clear", "[rb_get_free_after_clear]") {
+    lockfree::spsc::RingBuf<uint64_t, 1024U> rb;
+    const std::array<uint64_t, 512> test_data = {0xE5U};
+
+    rb.Write(test_data);
+    rb.Clear();
+
+    REQUIRE(rb.GetFree() == 1024U -1U);
+}
+
+TEST_CASE("Try to read after clear", "[rb_read_clear]") {
+    lockfree::spsc::RingBuf<uint64_t, 512U> rb;
+
+    uint64_t test_data_read[320] = {0};
+    bool const read_success = rb.Read(
+        test_data_read, sizeof(test_data_read) / sizeof(test_data_read[0]));
+
+    REQUIRE(!read_success);
+}
+
```

Comment:
I really don't like the first option, and regarding the second I'm not sure to understand why do we need two methods. I'm thinking in something like
```
template<typename T, size_t size>
void RingBuf<T, size>::Clear() {
    _r.store(0, std::memory_order_release);
    _w.store(0, std::memory_order_release);
}
```

Reply:
This operation cannot be thread safe, as it changes both the read and write index. A SPSC Queue works with the assumption that the write index will only ever be changed by the producer and the read index only ever changed by the consumer.

A store with the release memory ordering pairs with a load of the same variable with acquire memory ordering in another thread and they ensure that everything that happens before the release store is visible before the acquire load.

With this in mind, let's take a look at a scenario where the consumer is clearing the buffer, while the producer is writing:

```cpp
template <typename T, size_t size> bool Queue<T, size>::Push(const T &element) {
    /*
       The full check needs to be performed using the next write index not to
       miss the case when the read index wrapped and write index is at the end
     */
    const size_t w = _w.load(std::memory_order_relaxed);
    size_t w_next = w + 1;
    if (w_next == size) {
        w_next = 0U;
    }

    /* Full check  */
    const size_t r = _r.load(std::memory_order_acquire);
    if (w_next == r) {
        return false;
    }

    /* Place the element */
    _data[w] = element;

    /* Store the next write index */
    _w.store(w_next, std::memory_order_release);
    return true;
}
```

The load and store on `_r` ensure that this part will be visible (happens before) to the writer trying to clear the buffer:
```cpp
 const size_t w = _w.load(std::memory_order_relaxed);
    size_t w_next = w + 1;
    if (w_next == size) {
        w_next = 0U;
    }
```

As this is the only guarantee that we get, it can happen that the writer will think that `_w` is a completely different value from 0 we are setting in the consumer and later down the line we will replace it with `w_next`.

Even if we were to replace `const size_t w = _w.load(std::memory_order_relaxed);` with `const size_t w = _w.load(std::memory_order_acquire);`, this is not going to work as within the producer and the consumer the ordering of modifications to the indexes is going to be the same, but it would need to be different (and I'm not sure it would work even then, would need to check).

Reply:
As for the second approach, it would be thread safe because the consumer could change the read index to match the write index, whatever it is, and vice versa for the producer.

Reply:
After analyzing this a bit I've found that even `_r = _w`, or doing `Read()` in a loop might not be enough, because we can't know if there is a write in progress and, `_w` will change immedeately after. 

The reason I think this is a problem is because usually you want to flush when some assumption has changed and buffered data might not be valid anymore.

Reply:
A possible solution would be to use atomic flags for signalling from the consumer to the producer and vice versa. This would add overhead for every single operation though.